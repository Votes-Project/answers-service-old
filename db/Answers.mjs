// Generated by ReScript, PLEASE EDIT WITH CARE

import * as Answers__edgeql from "./__generated__/Answers__edgeql.mjs";

function allAnswers(client) {
  var query = Answers__edgeql.AllAnswers.query;
  return query(client);
}

var addAnswer = Answers__edgeql.AddAnswer.transaction;

var removeAnswer = Answers__edgeql.RemoveAnswer.transaction;

var AddAnswer;

var RemoveAnswer;

export {
  allAnswers ,
  AddAnswer ,
  addAnswer ,
  RemoveAnswer ,
  removeAnswer ,
}
/* Answers__edgeql Not a pure module */
