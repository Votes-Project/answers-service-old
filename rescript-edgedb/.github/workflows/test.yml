name: Test

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

jobs:
  build:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.node-version }}

      - name: Setup Bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest

      - name: Install dependencies
        run: npm ci

      - name: Build project
        run: npm run build

      - name: Run tests
        run: npm test

      - name: Run bundled CLI
        run: ./dist/Cli.js --help

      - name: Setup EdgeDB
        uses: edgedb/setup-edgedb@v1
        with:
          instance-name: "dbTestProject"

      - name: Init EdgDB
        working-directory: ./dbTestProject
        run: edgedb project init --non-interactive

      - name: Setup and build DB test project
        working-directory: ./dbTestProject
        run: npm ci && npm run build:rescript && npm run build:edgedb

      - name: Run DB tests
        working-directory: ./dbTestProject
        run: npm test
