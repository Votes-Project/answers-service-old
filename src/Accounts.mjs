// Generated by ReScript, PLEASE EDIT WITH CARE

import * as Accounts__edgeql from "./__generated__/Accounts__edgeql.mjs";

var addAccount = Accounts__edgeql.AddAccount.transaction;

var removeAccount = Accounts__edgeql.RemoveAccount.transaction;

var AddAccount;

var RemoveAccount;

export {
  AddAccount ,
  addAccount ,
  RemoveAccount ,
  removeAccount ,
}
/* Accounts__edgeql Not a pure module */
