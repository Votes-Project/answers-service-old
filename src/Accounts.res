module AddAccount = %edgeql(`
    # @name AddAccount
    insert Account {
      context_id := <str>$context_id,
    }
`)

let addAccount = AddAccount.transaction

module RemoveAccount = %edgeql(`
    # @name RemoveAccount
    delete Account
    filter .id = <uuid>$id
`)

let removeAccount = RemoveAccount.transaction
