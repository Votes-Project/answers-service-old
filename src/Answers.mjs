// Generated by ReScript, PLEASE EDIT WITH CARE

import * as Answers__edgeql from "./__generated__/Answers__edgeql.mjs";

function allAnswers(client) {
  var query = Answers__edgeql.AllAnswers.query;
  return query(client);
}

var addAnswer = Answers__edgeql.AddAnswer.transaction;

var AddAnswer;

export {
  allAnswers ,
  AddAnswer ,
  addAnswer ,
}
/* Answers__edgeql Not a pure module */
