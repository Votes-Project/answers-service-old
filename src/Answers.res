let allAnswers = client => {
  let query = %edgeql(`
    # @name allAnswers
    select Answer {
      id,
      answer_num_by_user,
      user: {
        context_id
      }
    } order by .answer_num_by_user;
`)
  client->query
}

module AddAnswer = %edgeql(`
    # @name AddAnswer
    insert Answer {
        answer_num_by_user := <int16>$answer_num_by_user,
        day := <str>$day,
        user := (
            select Account filter .context_id = <str>$context_id
        ),
        option := (
            select Option filter .id = <uuid>$id
        )
    }
`)

let addAnswer = AddAnswer.transaction
