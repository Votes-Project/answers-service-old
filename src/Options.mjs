// Generated by ReScript, PLEASE EDIT WITH CARE

import * as Options__edgeql from "./__generated__/Options__edgeql.mjs";

var addOption = Options__edgeql.AddOptions.transaction;

var removeOption = Options__edgeql.RemoveOption.transaction;

var AddOptions;

var RemoveOption;

export {
  AddOptions ,
  addOption ,
  RemoveOption ,
  removeOption ,
}
/* Options__edgeql Not a pure module */
