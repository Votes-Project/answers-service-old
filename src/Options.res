module AddOptions = %edgeql(`
    # @name AddOptions
    insert Option {
        option := <str>$option,
        details := <optional str>$details,
        question := (
          select Question filter .id = <uuid>$question_id
        )
    }
`)

let addOption = AddOptions.transaction

module RemoveOption = %edgeql(`
    # @name RemoveOption
    delete Option
    filter .id = <uuid>$id
`)
let removeOption = RemoveOption.transaction
