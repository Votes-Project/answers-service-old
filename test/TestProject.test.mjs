// Generated by ReScript, PLEASE EDIT WITH CARE

import * as Edgedb from "edgedb";
import * as Answers from "../src/Answers.mjs";
import * as Options from "../src/Options.mjs";
import * as Accounts from "../src/Accounts.mjs";
import * as Buntest from "bun:test";
import * as Questions from "../src/Questions.mjs";
import * as Core__Option from "@rescript/core/src/Core__Option.mjs";
import * as Core__Result from "@rescript/core/src/Core__Result.mjs";

var client = Edgedb.createClient(undefined);

Buntest.afterAll(async function () {
      return await client.close();
    });

function removeIds(key, value) {
  if (key === "id" && !(!Array.isArray(value) && (value === null || typeof value !== "object") && typeof value !== "number" && typeof value !== "string" && typeof value !== "boolean" || typeof value !== "string")) {
    return "<id>";
  } else {
    return value;
  }
}

Buntest.describe("fetching data", (function () {
        Buntest.test("fetching answers", (async function () {
                var answers = await Answers.allAnswers(client);
                var answers$1 = Core__Option.getWithDefault(JSON.stringify(answers, removeIds, 2), "");
                Buntest.expect(answers$1).toMatchSnapshot();
              }));
      }));

Buntest.describe("add and remove", (function () {
        Buntest.test("add and remove account transaction", (async function () {
                var res = await client.transaction(async function (transaction) {
                      return await Accounts.addAccount(transaction, {
                                  context_id: "0"
                                });
                    });
                var tmp;
                if (res.TAG === "Ok") {
                  var id = res._0.id;
                  await client.transaction(async function (transaction) {
                        return await Accounts.removeAccount(transaction, {
                                    id: id
                                  }, undefined);
                      });
                  tmp = id.length > 2;
                } else {
                  tmp = false;
                }
                Buntest.expect(tmp).toBe(true);
              }));
        Buntest.test("add and remove question", (async function () {
                var res = await client.transaction(async function (transaction) {
                      return await Questions.addQuestion(transaction, {
                                  question: "What is the meaning of life?"
                                });
                    });
                var tmp;
                if (res.TAG === "Ok") {
                  var id = res._0.id;
                  await client.transaction(async function (transaction) {
                        return await Questions.removeQuestion(transaction, {
                                    id: id
                                  }, undefined);
                      });
                  tmp = id.length > 2;
                } else {
                  tmp = false;
                }
                Buntest.expect(tmp).toBe(true);
              }));
        Buntest.test("add and remove option", (async function () {
                var questionId = Core__Result.mapWithDefault(await client.transaction(async function (transaction) {
                          return await Questions.addQuestion(transaction, {
                                      question: "What is the meaning of life?"
                                    });
                        }), "", (function (param) {
                        return param.id;
                      }));
                var res = await client.transaction(async function (transaction) {
                      return await Options.addOption(transaction, {
                                  option: "42",
                                  question_id: questionId
                                });
                    });
                var tmp;
                if (res.TAG === "Ok") {
                  var id = res._0.id;
                  await client.transaction(async function (transaction) {
                        return [
                                await Options.removeOption(transaction, {
                                      id: id
                                    }, undefined),
                                await Questions.removeQuestion(transaction, {
                                      id: questionId
                                    }, undefined)
                              ];
                      });
                  tmp = id.length > 2;
                } else {
                  tmp = false;
                }
                Buntest.expect(tmp).toBe(true);
              }));
        Buntest.test("add and remove answer", (async function () {
                var voterId = Core__Result.mapWithDefault(await client.transaction(async function (transaction) {
                          return await Accounts.addAccount(transaction, {
                                      context_id: "0"
                                    });
                        }), "", (function (param) {
                        return param.id;
                      }));
                var questionId = Core__Result.mapWithDefault(await client.transaction(async function (transaction) {
                          return await Questions.addQuestion(transaction, {
                                      question: "What is the meaning of life?"
                                    });
                        }), "", (function (param) {
                        return param.id;
                      }));
                var optionId = Core__Result.mapWithDefault(await client.transaction(async function (transaction) {
                          return await Options.addOption(transaction, {
                                      option: "42",
                                      question_id: questionId
                                    });
                        }), "", (function (param) {
                        return param.id;
                      }));
                var res = await client.transaction(async function (transaction) {
                      return await Answers.addAnswer(transaction, {
                                  answer_num_by_voter: 0,
                                  day: "0",
                                  voter_id: voterId,
                                  option_id: optionId
                                });
                    });
                var tmp;
                if (res.TAG === "Ok") {
                  var id = res._0.id;
                  await client.transaction(async function (transaction) {
                        return [
                                await Answers.removeAnswer(transaction, {
                                      id: id
                                    }, undefined),
                                await Options.removeOption(transaction, {
                                      id: optionId
                                    }, undefined),
                                await Questions.removeQuestion(transaction, {
                                      id: questionId
                                    }, undefined),
                                await Accounts.removeAccount(transaction, {
                                      id: voterId
                                    }, undefined)
                              ];
                      });
                  tmp = id.length > 2;
                } else {
                  tmp = false;
                }
                Buntest.expect(tmp).toBe(true);
              }));
      }));

Buntest.test("run unused selections CLI", (function () {
        var res = Bun.spawnSync([
              "npx",
              "rescript-edgedb",
              "unused-selections"
            ]);
        Buntest.expect(res.stdout.toString(undefined)).toMatchSnapshot();
      }));

export {
  client ,
  removeIds ,
}
/* client Not a pure module */
