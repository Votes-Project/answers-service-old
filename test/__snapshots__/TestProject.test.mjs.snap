// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`fetching data fetching answers 1`] = `"[]"`;

exports[`run unused selections CLI 1`] = `
"Analyzing project... (this might take a while)

[31mâœ˜[0m Found 4 unused selections.

File "Answers":
   In query "AllAnswers":
    - user: context_id
    - id
    - answer_num_by_user
    - user
"
`;
